codegen/plugin-gen/main.go

```diff
@@ -0,0 +1,108 @@
+// plugin-gen is responsible for auto-generating plugin code, specifically go
+// types based on json schema, for both payload and config
+package main
+
+import (
+	"fmt"
+	"go/build"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/docopt/docopt-go"
+	"github.com/taskcluster/jsonschema2go"
+)
+
+var (
+	version = "plugin-gen 1.0.0"
+	usage   = `
+plugin-gen
+plugin-gen is a tool for generating source code for a plugin, based on two json schema files
+(stored as either json or yaml) for its config settings and its portion of the task payload.
+As well as creating go types that the payload and config can be unmarshalled into, it also
+generates the PayloadSchema() and ConfigSchema() method implementations for the plugin.
+
+  Usage:
+    plugin-gen -c CONFIG-SCHEMA-FILE -p PAYLOAD-SCHEMA-FILE -o GO-OUTPUT-FILE
```

Comment:
I would propose the following:
 1) this tool be called `go-composite-schema`
 2) Commandline be on the form: `go-composite-schema [--required] <property> <input.yml> <output.go>`
 3) `<input.yml>` must be a JSON schema
 4) `<output.go>` will be a file as follows:
```go
type <schema-title> struct {
  .... // Whatever is generated from the schema
};

var <schema-title>Schema = func() runtime.CompositeSchema {
  schema, err := NewCompositeSchema('<property>', `
    <contents of input.yml as json>
  `, <true if --required>, func() interface{} { return &<schema-title>{}; });
  if err != nil {
    panic(err)
  }
  return schema;
}()
```

Maybe instead of using `<schema-title>` we could define a property called `typeName` on the schema, which jsonschema2go could use if present. That way we can name the type something sane, without compromising our documentation.

@petemoore, I would even be open to allowing you to add `typeName` as a top-level property to all our schemas in all TC components... Granted that would obviously be a lot of work to do :)

With this tool, the PluginProvider implementation becomes very simple, you just return `<schema-title>Schema` from the `ConfigSchema()` method...
I don't advice that we auto generate that method, because it's very easy to write... And this way we have a generation tool that ensures all of our `CompositeSchema`s are sane... And we can use it where ever we have CompositeSchemas. Whather for engine/plugin payload, or engine/plugins config, or config for some completely different sub-system we make in the future.

-----

On-topic, I suggest that we focus our energy on something other than this... but if you want to finish it up, this is a direction that would keep it simple, relevant and ensure that it can be widely used throughout our codebase. By focusing on CompositeSchema rather than plugins and engines.

Note: main value here is obviously that we can write in YAML, we can validate the schema again the meta-schema upfront, and we probably should do that :)